Review

Jinyang-Zou opened the pull request.

Jinyang-Zou committed:
init commit

Jinyang-Zou committed:
refactor(Reference): error management

modify validate_ref method, now can check for both create and update
add expected_params for ListReferences, if the user enters the wrong parameter, an error will occur
add customized api exceptions

Jinyang-Zou committed:
refactor(StatisticsSerializer): fixed stock check condition

Jinyang-Zou committed:
test: add unittests for reference and statistique

Enjoyted commented on bar/settings.py:
```diff
@@ -0,0 +1,138 @@
+    'django.contrib.staticfiles',
+    'rest_framework',
+    'rest_framework.authtoken',
+    'bar_app.apps.BarAppConfig',
```
it's a weird way to import your app. try only importing `bar_app` see if it works. django should implicitly execute the config

Enjoyted commented on bar_app/models.py:
```diff
@@ -0,0 +1,80 @@
+    """
+    Model representing a reference of beer.
+    """
+    id = models.AutoField(primary_key=True)
```
You are overriding django's default behaviour by doing this. Django does that already, but with a BigAutoField instead!
Who knows if your bar meets high success ? ;)
 (don't need to remove)

Enjoyted commented on bar_app/models.py:
```diff
@@ -0,0 +1,80 @@
+    """
+    Model representing a bar.
+    """
+    id = models.AutoField(primary_key=True)
```
useless

Enjoyted commented on bar_app/models.py:
```diff
@@ -0,0 +1,80 @@
+    """
+    Model representing an item in an order.
+    """
+    reference = models.ForeignKey(Reference, on_delete=models.CASCADE)
```
Not sure you want to CASCADE in this case. If you delete a product, you delete all related OrderItems related to it ?
Not sure the accounting service will be happy :)

Jinyang-Zou replied:
If the business requires avoiding cascading deletions, this approach may result in fewer traces left and help maintain data consistency. But also, I can implement alternative business logic if needed.

Enjoyted commented on bar_app/models.py:
```diff
@@ -0,0 +1,80 @@
+    Model representing an order.
+    """
+    id = models.AutoField(primary_key=True)
+    bar = models.ForeignKey(Bar, on_delete=models.CASCADE)
```
same comment as above

Jinyang-Zou replied:
If the related bar has been deleted, should we keep the order or retain a trace? Perhaps I should have asked a related question earlier

Enjoyted replied:
This is more of a product related problem, and the subject did not specify product related problems. As is, your code is not wrong.

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+from rest_framework import serializers
+from rest_framework.exceptions import ValidationError
+from django.contrib.auth.models import User
+from .models import *
```
I wouldn't recommend to blindly import everything from a module in general. Leave it as is

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+from django.contrib.auth.models import User
+from .models import *
+
+class UserSerializer(serializers.ModelSerializer):
```
this is not needed in the first place. Why is it here ?

Jinyang-Zou replied:
Initially I just wanted to get the user's token, after that I thought it wouldn't be a bad idea to create a view for managing user registrations and logins, so I created UserSerializer there for user's fields.

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+        for stock in stocks:
+            if stock.stock > 0:
+                return "available"
+        return "outofstock"
```
Although this answer initial demand, this isn't the most optimized approach.
This function will be executed at each instance serialization, meaning you are hitting the database everytime the serializer gets called.
Could it be done by annotating queryset on view instead ?

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+        """
+        if not value.strip():
+            raise serializers.ValidationError("Name field cannot be empty.")
+        return value
```
this is useless and already done by DRF. can you explain me why/how ?

Jinyang-Zou replied:
I'm so used to developing my own logic that I don't pay attention to the fact that DRF has handled these, same for the view

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+        if self.instance:
+            if not Bar.objects.filter(id=value).exists():
+                raise serializers.ValidationError("Invalid bar id.")
+        return value
```
this is useless. you can't specify an ID. it's an auto field.

Jinyang-Zou replied:
I wanted check if the user trying to update a bar that doesn't exist via the endpoint /api/bars/update/{bar_id}, now I figured out that DRF already handled all

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+            raise serializers.ValidationError("This reference already exists.")
+        return value
+
+    def validate_name(self, value):
```
this is useless and already done by DRF. can you explain me why/how ?

Jinyang-Zou replied:
Same for the previous comment

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+        fields = "__all__"
+        read_only_fields = ["id"]
+
+    def validate_ref(self, value):
```
this is useless and already done by DRF. can you explain me why/how ?

Jinyang-Zou replied:
Same for the previous comment

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+    all_stocks = serializers.SerializerMethodField()
+    miss_at_least_one = serializers.SerializerMethodField()
+
+    def get_all_stocks(self):
```
same remark as the availability. this should be done at queryset level.

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+        bars = Bar.objects.all()
+
+        for bar in bars:
+            stocks = Stock.objects.filter(bar=bar).all()
```
you should (virtually) never do a database hit inside a loop. use only django ORM

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+        bars = Bar.objects.all()
+
+        for bar in bars:
+            stocks = Stock.objects.filter(bar=bar).all()
```
you should (virtually) never do a database hit inside a loop. use only django ORM

Enjoyted commented on bar_app/serializers.py:
```diff
@@ -0,0 +1,169 @@
+            "bars": list(set(bars_with_all_stocks))
+        }
+
+    def get_miss_at_least_one(self):
```
same remark as the availability. this should be done at queryset level.

Enjoyted commented on bar_app/tests/reference/test_reference_model.py:
```diff
@@ -0,0 +1,81 @@
+    @classmethod
+    def setUpTestData(cls):
+        Reference.objects.create(ref="REF001", name="name 1", description="description 1")
+        Reference.objects.create(ref="REF002", name="name 2")
```
you should define those as class attributes and reuse them in all your tests. you would save at least one line of code in each test
(like you are doing in test_reference_view)

Enjoyted commented on bar_app/tests/reference/test_reference_model.py:
```diff
@@ -0,0 +1,81 @@
+
+    def test_ref_unique(self):
+        with self.assertRaises(Exception):
+            Reference.objects.create(ref="REF001", name="name", description="description")
```
you could reuse the variable (class attributes) to make your tests more dynamic and less hardcoded

Enjoyted replied:
also, this test is useless, because this is initially handled by DRF/django. You are making this test because you created useless validation methods.
You are basically testing the framework, and the people from django have done that for you a while back ;)

Enjoyted commented on bar_app/tests/reference/test_reference_model.py:
```diff
@@ -0,0 +1,81 @@
+
+    def test_verbose_name_plural(self):
+        verbose_name_plural = Reference._meta.verbose_name_plural
+        self.assertEqual(verbose_name_plural, "References")
```
this file could be removed, it contains nothing really worth. the only thing worth testing in your models are the __str__ functions

Enjoyted commented on bar_app/tests/statistique/test_statistique_serializer.py:
```diff
@@ -0,0 +1,51 @@
+from django.test import TestCase
```
file has french name, keep consistency

Enjoyted commented on bar_app/tests/statistique/test_statistique_serializer.py:
```diff
@@ -0,0 +1,51 @@
+        Stock.objects.create(reference=cls.reference3, bar=cls.bar1, stock=8)
+        Stock.objects.create(reference=cls.reference1, bar=cls.bar2, stock=5)
+        Stock.objects.create(reference=cls.reference2, bar=cls.bar2, stock=8)
+        Stock.objects.create(reference=cls.reference3, bar=cls.bar2, stock=0)
```
I wish you created fixtures. If you have time, you could take a look at these

Enjoyted commented on bar_app/tests/statistique/test_statistique_vieiw.py:
```diff
@@ -0,0 +1,51 @@
+from rest_framework.test import APITestCase
```
file name has french, and has a typo. not a big deal, but I wouldn't merge that

Enjoyted commented on bar_app/views/auth/auth_view.py:
```diff
@@ -0,0 +1,43 @@
+
+    def post(self, request, *args, **kwargs):
+        request.user.auth_token.delete()
+        return Response({"detail": "User logged out successfully."})
```
why this view?

Jinyang-Zou replied:
It's a complement, this will alow me to set permissions for other views to access it, using this view to retrieve the token after user logged in. I'll also revoke the token when the user logged out. To ensures that anonymous users cannot access these endpoints.

Enjoyted replied:
It would have been good to explain why you added things that were not initially asked in the project. Think of that as a way to sell me your project.

It's hard for me to understand why you did something, did you think it was needed in order to make the project function ? Was it a bonus to make it better ?

Thank you for the explanation

Jinyang-Zou replied:
I assumed that any endpoints not specifically restricted can be accessed without authentication, as the client didn't specify in the project. Through this view, I can set permissions to limit access to specific endpoints only to logged-in users or those with staff privileges. This helps prevent unauthorized access to sensitive information. I can also create custom permission classes to have view level access control rather than just object level checks by default. These features enhance both the functionality and security of the project. What do you think ?

Jinyang-Zou replied:
I actually wanted to develop a JWT login system so it would be better, but the project didn't ask for anything about login system, so I just used the simplest token, but of course django can use username and password authentication methods too!

Enjoyted commented on bar_app/views/bars/bar_view.py:
```diff
@@ -0,0 +1,70 @@
+        if isinstance(request.data, dict):
+            serializer = self.get_serializer(data=request.data)
+        elif isinstance(request.data, list):
+            serializer = self.get_serializer(data=request.data, many=True)
```
what is this piece of code supposed to handle ? You can create batches of bars ? 😕

Jinyang-Zou replied:
Yes it is, create one bar or multiple bars (same for reference). This could be an option maybe?

Enjoyted replied:
it changes quite a lot the initial API asked. but keep it as is

Enjoyted reviewed (commented):
Hello Zou!
Thank you for doing the test :)

Can you take into account the threads I made, and either do modifications to the code, or reply to me (disagree, or tell me how you would fix)

There are two things I didn't say in the review that I will say here:
- Views : I will ask you to redo one endpoint (the Bar) completely, using ModelViewSet. You are writing a lot of code that is handle natively by ModelViewSet. Also, it should have been obvious since you (correctly) used ModelSerializers!
- Urls : Look into routers. Your views are too hand crafted. Also, less route files please.

Overall, you should aim to write less code, not because it's always better, but because it means you understand what the framework brings to the table. A lot of things you do are handled natively if you read a bit more the documentation.

Overall, there are good things, so nothing to be ashamed of. Time to make this mergeable now 🚀 

Thank you

Jinyang-Zou committed:
refactor: modifications after code review

rename statistique to statistics for related files or folders
modify the way to import bar_app in setting.py
apply new method for bars route management
remove useless id field from Bar model
remove useless bars urls file that already handled by router
remove useless test_reference_model.py

Jinyang-Zou commented:
> Hello Zou! Thank you for doing the test :)
> 
> Can you take into account the threads I made, and either do modifications to the code, or reply to me (disagree, or tell me how you would fix)
> 
> There are two things I didn't say in the review that I will say here:
> 
> * Views : I will ask you to redo one endpoint (the Bar) completely, using ModelViewSet. You are writing a lot of code that is handle natively by ModelViewSet. Also, it should have been obvious since you (correctly) used ModelSerializers!
> * Urls : Look into routers. Your views are too hand crafted. Also, less route files please.
> 
> Overall, you should aim to write less code, not because it's always better, but because it means you understand what the framework brings to the table. A lot of things you do are handled natively if you read a bit more the documentation.
> 
> Overall, there are good things, so nothing to be ashamed of. Time to make this mergeable now 🚀
> 
> Thank you


Hello Teddy,

Thank you very much for the code review, it helps me understand better the framework!

I refactored bar view and its router, but i left other routers like before if that's okay with you ？Because router requires viewset, but I've used APIView for the most of endpoints which require more modifications..

I've also created a db_fixture class by the way so that all the tests can share the same test data, not sure if what I did is the right thing to do, but I tried to fix other bugs in the comments, feel free to give me feedback!

Jinyang-Zou committed:
refactor: unittest, bar model, router

apply new router method for bar urls
modify bar view from generic ApiView to ModelViewSet
refactor get_availability, get_all_stocks, get_miss_at_least_one methods and apply queryset method
remove useless id field from Bar model
remove useless params check for ListreRerences

apply db_fixture for all tests
